solidity/contracts/bridge/library/Fraud.sol

```diff
@@ -0,0 +1,396 @@
+
+        emit FraudChallengeSubmitted(walletPubKeyHash, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function defeatFraudChallenge(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes memory preimage,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bool witness,
+        address treasury,
+        mapping(uint256 => Bridge.DepositRequest) storage deposits,
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) external {
+        bytes32 sighash = preimage.hash256();
+
+        uint256 challengeKey = uint256(
+            keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge closed");
+
+        // Ensure SIGHASH_ALL type was used during signing, which is represented
+        // by type value `1`.
+        require(extractSighashType(preimage) == 1, "Wrong sighash type");
+
+        if (witness) {
+            verifyWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        } else {
+            verifyNonWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        }
+        // If we passed the preimage verification, the wallet has successfully
+        // defended the fraud challenge.
+
+        // Send the ether deposited by the challenger to the treasury
+        /* solhint-disable avoid-low-level-calls */
+        // slither-disable-next-line low-level-calls
+        (bool success, ) = treasury.call{value: challenge.depositAmount}("");
+        require(success, "Failed to send Ether");
+        /* solhint-enable avoid-low-level-calls */
+
+        // TODO: Reward the caller for defeating the challenge successfully.
+        challenge.closed = true;
+
+        bytes memory compressedWalletPublicKey = walletPublicKey
+            .compressPublicKey();
+        bytes20 walletPubKeyHash = bytes20(compressedWalletPublicKey.hash160());
+
+        emit FraudChallengeDefeated(walletPubKeyHash, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function notifyChallengeTimeout(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        uint256 challengeKey = uint256(
+            keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge is closed");
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - challenge.reportedAt >=
+                self.challengeDefendTimeout,
```

Comment:
Maybe `block.timestamp >= challenge.reportedAt + self.challengeDefendTimeout`? It's the same but a bit easier to grasp.

Reply:
Done.